### PR TITLE
Adds ability to exclude certain files from app directories

### PR DESCRIPTION
**Pull request by muzfuz**
I have found it useful in testing against a large Rails application to exclude certain files from method tracing altogether.

This PR introduces a new `AppFiles` module that holds the logic and config for doing this.  

I considered calling this the `IgnoredFiles` module, but it seems to make sense to be a bit more generic about it, as there might also be situations where someone would want to _include_ a file that is not found in the `application_directories`, in which case this class would be easy to extend with that additional behaviour.

Happy to consider any additional changes.

### CODE REVIEW

**Comment by muzfuz** [comment]
One thing I don't love at the moment is that there are now two classes that have a filename cache.  Seems needless extra overhead.

We could theoretically unify the two classes into one, since they both do very similar things.  I don't particularly love that idea, but it may be preferable to the current alternative.

**Comment by markburns** [comment]
Just had a realisation of what we need 

We're missing a config block at the moment '.setup!' deals with config and starting the method logging and it should only do one of those things 

We need something like:

Delfos.config do |c|
  c.application_directories << "app"
  c.ignore_files << "file.rb"
end

Then we can keep most of this implementation as it is.
Then we should rename 'Delfos.setup!' to 'Delfos.start!'

**Comment by muzfuz** [comment]
@markburns I was thinking the same thing.  We're now getting to a point where he have so much configuration that a config block makes sense.  I'm happy to write that and include it in this PR.

**Comment by markburns** [comment]
That'd be great

**Comment by muzfuz** [comment]
@markburns done.  Haven't updated readme yet, please have a look and feel free to make any changes you like.

**Comment by muzfuz** [comment]
Ah, local Rubocop did not catch that codeclimate issue.  I'll fix.

**Comment by muzfuz** [comment]
Ok I've pushed up a few more changes.  Gonna clock out for the day, will pick this up in the morning.

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@
 /*.bundle
 .agignore
 tags
+delfos_query_parameters.json
```
Comment (markburns) [inline .gitignore]: I'll make this output into `./tmp` by default

**Comment by muzfuz** [comment]
Morning @markburns - happy to make any adjustments this morning.  Be good to hear your thoughts.

I don't have permissions to tell Codeclimate to ignore that Rubocop issue, so I'll have to rely on you to do that please :)

**Comment by markburns** [comment]
Hey I've done a bunch of changes but haven't got the build to green yet.

I've tried to simplify the whole tons of state/class instance variables, but ironically there's stuff persisting in between specs.

I can push now and you can have a look. Or rework it?
I think the intention should be clearer from looking at the code.

**Comment by muzfuz** [comment]
Ah cool.  Yeah sure, happy to take a look this morning, unless you want a bit more time to poke at it :)

**Comment by markburns** [comment]
Yeah not sure I'll have much more time to look at it today. 

If it's too painful getting all of these changes in/fixing the build I'm happy to have a slimmed down version of this PR.

I'd want to have the `include` and `exclude` stuff. 

Basically, there was too much going on in `Config` and `Setup` and by trying to move most of the class instance variables into the config instance, we can avoid a lot of the clean up steps between tests.

I think the problem is in the `MethodTrace` spec, but I haven't seen where the config is getting re-setup again.

It's a bit slow going because it freezes up the neo4j server with tons of open transactions/batches. So if you find it hanging you can restart your server.

File: lib/delfos/config/inclusion.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        append_to(files, "@#{type}_files")
+        append_to(directories, "@#{type}_directories")
+      end
+
+      def append_to(paths, ivar_name)
+        ivar = instance_variable_get(ivar_name)
+        ivar += paths
+        ivar.uniq!
+        instance_variable_set(ivar_name, ivar)
```
Comment (markburns) [inline lib/delfos/config/inclusion.rb]: I started to DRY up this code, but not convinced it's easier to work with. 
Don't mind if this gets changed to something a bit more verbose. Or maybe there's a different nicer way to do it

File: lib/delfos/config_spec.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      it "replaces the existing paths" do
+        subject.exclude = "app"
+
+        expect(subject.excluded_directories).to eq [app_path]
+      end
+    end
+  end
+end
+
+# Delfos.config do |c|
```
Comment (markburns) [inline lib/delfos/config_spec.rb]: need to delete these notes

File: lib/delfos/file_system/app_files.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+require_relative "file_cache"
+
+module Delfos
+  module FileSystem
+    class AppFiles
+      include FileCache
+
+      def initialize(included, excluded)
```
Comment (markburns) [inline lib/delfos/file_system/app_files.rb]: Started work on making this support included files too, but it isn't finished.

**Comment by muzfuz** [comment]
```ruby
      def setup_trace_point(type)
        TracePoint.new(type) do |tp|
          next unless Delfos.include_file?(tp.path)
          yield tp
        end
      end
```

`include_file?` now creates an instance of `Config` if it doesn't already exist.  This is why these instance variables keep being re-created.

**Comment by markburns** [comment]
Ah - good find. OK so maybe we should only make config objects explicitly created. 

```ruby
Delfos.configure{|c| # implicitly created }
Delfos.start! # implicitly created
Delfos.anything_else # raises error
```

**Comment by muzfuz** [comment]
I've pushed up a load of spec fixes just now.

If you run the `spec`, `call_stack`, `file_system`, `method_trace` folders individually, the specs pass.

The only thing still broken is the `neo4j` folder.  I am strongly suspicious that the specs in that folder are leaking, and causing all the other specs to go red.  (We get more failures when you run all the tests together).

**Comment by markburns** [comment]
With `rspec lib/neo4j --bisect` I found the importer spec was a cause. I've fixed that one

**Comment by markburns** [comment]
@muzfuz - fixed some specs

**Comment by muzfuz** [comment]
My local Neo4j DB is in use by another job at the moment, I'll pick this up as soon as that's done!

**Comment by muzfuz** [comment]
Ok so at the moment it looks like everything in `lib` runs except for the `method_trace_spec.rb`.

I cannot currently run the `spec` folder.  

It's a bit of a rabbit hole, but I'm happy to keep plugging away at this, as I think it's going to be very much worthwhile to have this feature go in.

**Comment by markburns** [comment]
Ineo allows you to run multiple instances of neo4j and set their ports etc

**Comment by markburns** [comment]
@muzfuz - OK done. 

Just FYI, just in case you are wondering, I'm not actually deleting those comments from the codacy bot. No idea what those messages are. Perhaps when I've dismissed something in the UI in the past? Seems a bit weird if codacy deletes the comments as me though.

**Comment by muzfuz** [comment]
Brilliant work @markburns.  And good to know about those comments haha!

**Pull request merged**